### ISSUE TITLE
Can't start app from ZIP artifact

### ISSUE THREAD

**Issue by pashvin** [electron-userland/electron-builder#4299]
* **Version**:  21.2.0

 6.x

* **Target**:  Mac/zip

Extract app from ZIP and try to start it, fails with "The application xyz can't be opened".

**Comment by jsgv** [comment]
I am experiencing the same thing. 

I can open an unsigned build, but not a zipped up signed build.

**Comment by jsgv** [comment]
Partially solved with [electron-notarize](https://github.com/electron/electron-notarize).

Due to Apple requirements on Catalina OS. https://developer.apple.com/news/?id=04102019a

**Comment by jookyboi** [comment]
@jsgv I'm using `electron_notarize` and it still fails for the same reason. Can you say more about how you got it to work?

**Comment by jookyboi** [comment]
Update: I've tried just compressing the `.app/` folder with Finder->Compress and the unzipped app works fine on Catalina. It has something to do with the 7z compressor used from `7zip-bin`. In the meantime, I'm trying to just use the native `zip` command to compress the folder.

**Comment by jsgv** [comment]
@jookyboi I mostly followed this article: 
https://medium.com/@TwitterArchiveEraser/notarize-electron-apps-7a5f988406db

Important details is creating the 'App-specific password'. Your regular apple credentials will not work.
https://support.apple.com/HT204397

**Comment by jookyboi** [comment]
@jsgv I don't have an issue notarizing the app (and neither does the issuer filer.) It has to do with using the `zip` build target for macOS. Something about the way 7z is compressing the app folder causes Catalina to reject opening the unzipped app.

**Comment by jsgv** [comment]
> @jsgv I don't have an issue notarizing the app (and neither does the issuer filer). It has to do with using the `zip` build target for macOS. Something about the way 7z is compressing the app folder causes Catalina to reject opening the unzipped app.

You are correct. My mistake.

**Comment by jookyboi** [comment]
Update: Running from my Electron app's root folder: 
`../../node_modules/7zip-bin-mac/7za a -mm=Deflate -r [path_to_zip] [MyApp.app/]`gives me a decently compressed zip file which still works fine on Catalina.

Can @develar look into removing some of the optional arguments for `7z` to get things working again? https://github.com/electron-userland/electron-builder/blob/b83f5801771585c18524c3cee74e59de5fa9ab30/packages/app-builder-lib/src/targets/archive.ts#L86

**Comment by jookyboi** [comment]
**Update:** 7zip has nothing to do with it. I tried with the max compression option from the command line and that still worked okay on Catalina. Something happening between zipping and final output is causing the issue.

**Comment by pashvin** [comment]
My application is notarized so there is no issue with it, having said that I don't know zip artifact waits for notarization to be finish. I can try to compare content from zip app vs dmg app and see that has any difference

**Comment by pashvin** [comment]
Ok I used following command to check app notarized or not from dmg and zip and it is confirmed that zip picked up app before notarization is done.

Command to check status of app:
codesign --test-requirement="=notarized" --verify --verbose <appname.app>

**Comment by jookyboi** [comment]
@pashvin Good find. So the issue is that the folder get zipped up before notarization is complete? It just feels strange as my build process resulted in a properly notarized zipped app in Mojave. w

**Comment by jookyboi** [comment]
Here is the workaround I'm using until the problem gets fixed: https://snippets.cacher.io/snippet/354a3eb7b0dcbe711383

**Comment by shahan312** [comment]
We've been experiencing the same issue and using a workaround similar to @jookyboi's.

**Comment by stevenroussey-privicy** [comment]
The last couple of days Apple has had issues with notarization, which if you do it by hand you would have noticed how long it took. Curious if the timing bug disappears when apple is running more smoothly today.

**Comment by pashvin** [comment]
No this is not a timing issue. This is always reproducible. I know notarization takes some time long and some time it is fast but I am tracking this issue from almost month and we have every day build. Every build has the same issue.

**Comment by jookyboi** [comment]
**Update:** Found the culprit: https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/targets/ArchiveTarget.ts#L65

After creating the zip, the default behavior for `ArchiveTarget` is to append the `.blockmap` to the file, then calculate the `blockmapsize` from that. This is in the name of doing differential updates.. though I've never seen `electron-updater` update this way for macOS.

I dug into `node_modules/app-builder-lib/out/targets/ArchiveTarget.js` and commented out this functionality:
```
if (this.isWriteUpdateInfo && format === "zip") {
  // updateInfo = await (0, _differentialUpdateInfoBuilder().appendBlockmap)(artifactPath);
}
```

The resulting `latest-mac.yml` will be missing the `blockmapsize` attribute but it doesn't seem to matter for me.

The resulting zip works fine on Catalina as does the auto-update functionality.

**Comment by pashvin** [comment]
Yes once app is signed by apple, any modification will make signature invalid. I think those item must be added (if really required) before starting app notarization. Looks like we have some solution and need to generate an official build with fix.

**Comment by jookyboi** [comment]
I think the correct course of action would be to append the `blockmapsize` attribute, create the zip, then notarize the zip file itself. According to the docs, this should be possible: https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution

**Comment by pashvin** [comment]
There is no need to notarize zip, if anything is added to .app, then it need to be added before it sent for notarization so correct fix is add everything to .app before notarization start and then make zip once notarization is done. remember ZIP is not complaining anything , it extracts fine but .app is not able to start.

**Comment by pashvin** [comment]
what is the next step to get the fix. I think we already proved that this is a real issue so to get the fix what is the next step? Need to assign to some developer? Waiting for fix for this issue almost 3 month.
I can't go back to older version because in that case i do not get any way to notarize app.

**Comment by semireg** [comment]
This is a very strange issue and I'll mention my findings because I haven't seen it mentioned here.

I had multiple Catalina users report "The application Label LIVE can't be opened." I downloaded and the app worked fine on Catalina. Url for reference is http://label.live/download. 

I finally screen-shared (via Zoom) with a user and sure enough, the download didn't work. However, they were using Chrome and the zip file was being decompressed by Archive Utility. I had the user download the same zip using Safari and presto, the decompressed app launched no problem. I have given this advice to other users with varying success. I believe that Safari users must have "Open safe files after downloading" checked. If not, then the Archive Utility is launched to unzip.

![2019-10-21 at 3 02 PM](https://user-images.githubusercontent.com/1819762/67239030-e4b2f480-f413-11e9-9dde-ba1a4443b81c.png)

In all of my own tests I was decompressing with Safari's built-in unzipper (whatever that is)... can anyone confirm this issue with their builds? I'm skeptical that I'm the only one to notice this behavior...

**Comment by ansgarschulte** [comment]
@semireg I can confirm, that Safari downloaded app works for me and Chrome downloaded and Finder unzipped does not.

**Comment by semireg** [comment]
Apple developer support helped me understand the issue. 🙌 

The problem is that Archive Utility (**only on Catalina**) fails to create the proper symlinks when unpacking, whereas Safari "open safe" does create the symlinks as expected. An example of this is `Contents/Frameworks/Electron Framework.framework/Versions/Current -> A`

But, before we can blame Archive Utility, consider that `unzip` (on both Mojave and Catalina) fails to list the contents of the zip:

```
unzip -l Label-LIVE-1.4.6-10406006.zip
Archive:  Label-LIVE-1.4.6-10406006.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of Label-LIVE-1.4.6-10406006.zip or
        Label-LIVE-1.4.6-10406006.zip.zip, and cannot find Label-LIVE-1.4.6-10406006.zip.ZIP, period.
```

Apple developer support recommended ditto to create the zip, but I'm not sure if this is our end goal... I have not looked into how to fix/work-around electron-builder. Looking for help with that! 😅 

`/usr/bin/ditto -c -k --keepParent "$APP_PATH" "$ZIP_PATH"`
Reference: https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution/customizing_the_notarization_workflow

I will file a bug with Apple regarding this regression in Catalina's Archive Utility.

**Comment by jookyboi** [comment]
@semireg Thanks so much for looking into the issue and touching base with Apple. If you'd like to hack around with the zip creation, I would take a look at https://github.com/electron-userland/electron-builder/blob/f0bf67c7d87e7af4e31b24f45172cf4ba3c2ca72/packages/app-builder-lib/src/targets/archive.ts#L147.

The function explicitly calls `7zip` to compress the `.app/` directory. You may be able to swap out a few things there.

**Comment by IDmedia** [comment]
I've also noticed this issue, but can only reproduce it when zipping my .app file and uploading it to Dropbox or another remote location. When a user then downloads the file and unpacks the zip file on Catalina they also get the "App is damaged" message. Just zipping and unzipping the file locally works without issues.

**Comment by mobitar** [comment]
> Here is the workaround I'm using until the problem gets fixed: https://snippets.cacher.io/snippet/354a3eb7b0dcbe711383

There seems to be a problem with this snippet. If you compress the .app file, then immediately try extracting and running on Catalina (without running the blockmap script), it will work fine. If however you compress the .app file, then run the `./node_modules/app-builder-bin/mac/app-builder blockmap -i dist/[YourApp]-${version}-mac.zip` command, this will _modify_ the .zip file, and this file will then extract properly on Catalina, but the inner .app file will not run.

You can see that running the blockmap command many times will yield different signatures each time.

**Comment by semireg** [comment]
See also https://github.com/develar/app-builder/issues/24

**Comment by vicatcu** [comment]
I don't know how to get this to work. I scp'd the App as a directory over to a linux machine and zipped it there, then sent that zip to the internet, downloaded it on my Mac where I built it, and when I launch it I get a dialog telling me "Your App Name" can't be opened because the identity of the developer cannot be  confirmed... Your security preferences all installation of only apps from the App Store and identified developers." Then I can circumvent that through my system settings, but that's not really an acceptable user experience for people downloading the app from my website. Should I just be uploading the DMG artifact in the dist folder instead of the zip or what?

**Comment by jthrilly** [comment]
> I don't know how to get this to work. I scp'd the App as a directory over to a linux machine and zipped it there, then sent that zip to the internet, downloaded it on my Mac where I built it, and when I launch it I get a dialog telling me "Your App Name" can't be opened because the identity of the developer cannot be confirmed... Your security preferences all installation of only apps from the App Store and identified developers." Then I can circumvent that through my system settings, but that's not really an acceptable user experience for people downloading the app from my website. Should I just be uploading the DMG artifact in the dist folder instead of the zip or what?

This sounds like your app isn't signed, which would be a different issue to this one.

**Comment by oskarhane** [comment]
@mobitar 
> The solution seems to be to have users download the .dmg from your website (as opposed to a zip), and have the app use the .zip for auto-update.

Do the zip (or .app inside it) need to be notarized as well as signed?
I have a working (signed + notarized) dmg now which I offer as downloads.
For the auto update I plan to offer the zipped app, but I currently only sign it, not notarize.
Would that put me in trouble when updating the app via `electron-updater`?

**Comment by pashvin** [comment]
This thread is diverted a lot from the original issue. This was created for the ZIP artifact comes out of the box by just simply adding option in package.json to generate ZIP and that file does not work in Catalina.
Can we fix or find fix for that first? I know if you host ZIP and any DMG file on web url and download from there , then there are other issue where Mac OS does not trust that package but in my case I am not hosting or downloading. I am directly unziping on build machine and trying to run and that does not work first place.

**Comment by mobitar** [comment]
Here's what finally worked for me:

First prepare .zip file:
1. Open `dist/mac`
2. Right click .app file and compress
3. Move .zip file to ~/Desktop (or wherever)

Then:
1. Clone https://github.com/develar/app-builder
2. `cd` into directory, then `go build`
3. Run `./app-builder blockmap -i ~/Desktop/myapp.app.zip -o ~/Desktop/throwaway.zip`
4. Copy the output `sha512` and `size` params into `yourapp/dist/latest-mac.yml`.
5. Upload `latest-mac.yml` and `~/Desktop/myapp.app.zip` to GitHub (or other CDN).

The important part here is the `-o` (output) arg in the app-builder command. This prevents app-builder from directly modifying the input file (which is what's causing the problems).

Both auto-update and direct zip extraction worked for me on Catalina with this method.

**Comment by pashvin** [comment]
So that is the fix (create a zip artifact before bundling to dmg. Use the same .app to zip) and we all know these manual steps works but we need to fix electron builder so no one need manual steps we are talking about these manual step over and over but it is time to fix the electron builder.

**Comment by sidneys** [comment]
> Here's what finally worked for me:
> 
> First prepare .zip file:
> 
>     1. Open `dist/mac`
> 
>     2. Right click .app file and compress
> 
>     3. Move .zip file to ~/Desktop (or wherever)
> 
> 
> Then:
> 
>     1. Clone https://github.com/develar/app-builder
> 
>     2. `cd` into directory, then `go build`
> 
>     3. Run `./app-builder blockmap -i ~/Desktop/myapp.app.zip -o ~/Desktop/throwaway.zip`
> 
>     4. Copy the output `sha512` and `size` params into `yourapp/dist/latest-mac.yml`.
> 
>     5. Upload `latest-mac.yml` and `~/Desktop/myapp.app.zip` to GitHub (or other CDN).
> 
> 
> The important part here is the `-o` (output) arg in the app-builder command. This prevents app-builder from directly modifying the input file (which is what's causing the problems).
> 
> Both auto-update and direct zip extraction worked for me on Catalina with this method.

This works as an interim solution - many thanks to you.

**Comment by LabhanshAgrawal** [comment]
Is there any update on this? The workaround is ok for now, but would prefer to have a proper fix.

**Comment by Aarbel** [comment]
Same issue there, looking for a good workaround

**Comment by matheuss** [comment]
We're seeing the same for https://hyper.is https://getkap.co 🤕

**Comment by jwheare** [comment]
This is blocking a release for https://github.com/irccloud/irccloud-desktop

**Comment by mobitar** [comment]
Hmm, the workaround I posted is no longer working for me for some reason. I can open the zipped .app file when it's generated locally. But if I upload to GitHub, then download it again, it won't open ("can’t be opened because Apple cannot check it for malicious software.")

**Comment by do-web** [comment]
> Here's what finally worked for me:
> 
> First prepare .zip file:
> 
> 1. Open `dist/mac`
> 2. Right click .app file and compress
> 3. Move .zip file to ~/Desktop (or wherever)
> 
> Then:
> 
> 1. Clone https://github.com/develar/app-builder
> 2. `cd` into directory, then `go build`
> 3. Run `./app-builder blockmap -i ~/Desktop/myapp.app.zip -o ~/Desktop/throwaway.zip`
> 4. Copy the output `sha512` and `size` params into `yourapp/dist/latest-mac.yml`.
> 5. Upload `latest-mac.yml` and `~/Desktop/myapp.app.zip` to GitHub (or other CDN).
> 
> The important part here is the `-o` (output) arg in the app-builder command. This prevents app-builder from directly modifying the input file (which is what's causing the problems).
> 
> Both auto-update and direct zip extraction worked for me on Catalina with this method.

Not working for me, i get an error "undefined"....
@mobitar any idea?

**Comment by jwheare** [comment]
Yes this is still very relevant. It requires changes to the zip step, or at the least some hooks for us to override that step with an appropriate workaround.

**Comment by philippkuehn** [comment]
Oh man these bots. Yeah it’s relevant as long it’s not fixed 👀

**Comment by tbuteler** [comment]
I tried all workarounds listed here and they didn't help in my case. No matter how I built the ZIP file, once I uploaded it to GitHub and downloaded it with Chrome, it would fail to open (inspecting the extracted application with `codesign` said that code signing didn't happen, but my release pipeline verifies both signature and notarization and confirmed all was well). Downloading the same ZIP file with Safari, on the other hand, would work as expected.

I'm on macOS 10.15.4 Catalina. Tried creating builds with `electron-builder` 22.5.1 and 21.2.0, both yielded the same results above.

I've retroactively added DMG releases on a couple of older versions of my app and can confirm that not only applications extracted from DMGs work fine, but auto-updating will work too, without any of the workarounds listed here (i.e. `electron-builder` updater can extract the app from that same ZIP that Chrome can't touch and upgrade locally without issues).

I'll just permanently move away from offering ZIP files for users to download on Mac, and use DMGs instead. Seems much more practical than waiting around for Apple to fix this issue. 

Thanks to all who posted their findings here, I really appreciate it.

**Comment by yanlee26** [comment]
> > Here's what finally worked for me:
> > First prepare .zip file:
> > ```
> > 1. Open `dist/mac`
> > 
> > 2. Right click .app file and compress
> > 
> > 3. Move .zip file to ~/Desktop (or wherever)
> > ```
> > 
> > 
> > Then:
> > ```
> > 1. Clone https://github.com/develar/app-builder
> > 
> > 2. `cd` into directory, then `go build`
> > 
> > 3. Run `./app-builder blockmap -i ~/Desktop/myapp.app.zip -o ~/Desktop/throwaway.zip`
> > 
> > 4. Copy the output `sha512` and `size` params into `yourapp/dist/latest-mac.yml`.
> > 
> > 5. Upload `latest-mac.yml` and `~/Desktop/myapp.app.zip` to GitHub (or other CDN).
> > ```
> > 
> > 
> > The important part here is the `-o` (output) arg in the app-builder command. This prevents app-builder from directly modifying the input file (which is what's causing the problems).
> > Both auto-update and direct zip extraction worked for me on Catalina with this method.
> 
> This works as an interim solution - many thanks to you.

still not work for me...
```
{
  constructor: 'Error',
  code: 2,
  domain: 'SQRLUpdaterErrorDomain',
  stack: 'Error: Could not locate update bundle for com.github.Electron within file:///Users/elliotyan/Library/Caches/com.github.Electron.ShipIt/update.nxcQidV/'

```

**Comment by jetlej** [comment]
Any updates?

**Comment by georgevs** [comment]
Electron-builder Version: 22.4.0
macOS version: Catalina 10.15.4
Node version: v12.16.1

It seems like the attributes of the symbolic links in the `Frameworks` subfolder in the app bundle in both `zip` and `tar.bz2` are lost and the files appear to the OS as regular files.

```
> ls -la ./dist/zip/Electron.app/Contents/Frameworks/Electron\ Framework.framework
-rw-rw-r--   1 georges  ...   35 May 15 18:49 Electron Framework
-rw-rw-r--   1 georges  ...  26 May 15 18:49 Libraries
-rw-rw-r--   1 georges  ...   26 May 15 18:49 Resources
drwxrwxr-x   4 georges  ...  128 May 15 18:49 Versions

> ls -la ./dist/mac/GoToMeeting-Electron.app/Contents/Frameworks/Electron\ Framework.framework
lrwxr-xr-x   1 georges  ...   35 May 15 18:49 Electron Framework -> Versions/Current/Electron Framework
lrwxr-xr-x   1 georges  ...   26 May 15 18:49 Libraries -> Versions/Current/Libraries
lrwxr-xr-x   1 georges  ...   26 May 15 18:49 Resources -> Versions/Current/Resources
drwxr-xr-x   4 georges  ...  128 May 15 18:49 Versions
```

**Comment by georgevs** [comment]
@MehediH I am not convinced that's the macOS only to blame.  Please note that if I manually tar/bzip2 the app in dist/mac everything looks and works fine with the generated tar.bz2 archive, and still the attributes are lost with the `electron-builder` generated tar.bz2:

```
... in ./MyApp/dist/mac ...
$ FN=MyApp ; /usr/bin/tar -cf ${FN}.tar ${FN}.app && /usr/bin/bzip2 -fz9 ${FN}.tar

... in ./MyApp/dist ...
$ tar tvf MyApp-1.0.0-mac.tar.bz2 | grep Electron\ Framework.framework/Libraries
-rwxrwxrwx  0 0      0          26 May 18 09:17 MyApp.app/Contents/Frameworks/Electron Framework.framework/Libraries

$ tar tvf mac/MyApp.tar.bz2 | grep Electron\ Framework.framework/Libraries
lrwxr-xr-x  0 georges ...      0 May 18 09:17 MyApp.app/Contents/Frameworks/Electron Framework.framework/Libraries -> Versions/Current/Libraries
```

**Comment by kamalbennani** [comment]
> This is a very strange issue and I'll mention my findings because I haven't seen it mentioned here.
> 
> I had multiple Catalina users report "The application Label LIVE can't be opened." I downloaded and the app worked fine on Catalina. URL for reference and public reproduction: http://label.live/download.
> 
> I finally screen-shared (via Zoom) with a user and sure enough, the download app didn't work. They were using Chrome and the zip file was being decompressed by Archive Utility. I had the user download the same zip using Safari and presto, the decompressed app launched no problem. I have given this advice to other users with varying success. I believe that Safari users must have "Open safe files after downloading" checked. If not, then the Archive Utility is launched to unzip - causing the issue.
> 
> ![2019-10-21 at 3 02 PM](https://user-images.githubusercontent.com/1819762/67239030-e4b2f480-f413-11e9-9dde-ba1a4443b81c.png)
> 
> In all of my own tests I was decompressing with Safari's built-in unzipper (whatever that is)... can anyone confirm this issue with their builds? I'm skeptical that I'm the only one to notice this behavior...

Thank you, man, for this finding 💪

**Comment by alexandreroche** [comment]
Would love a fix here!

**Comment by wavifu** [comment]
I created a small script to patch this on Mac Catalina.

https://gist.github.com/wavifu/7539e5bf84a715f6e728bd14479d477c

First, you need to let electron-builder build and release your app so that it generates latest-mac.yml. Then, run this script. This will re-pack the app using ditto, run app-builder command to retrieve has and blockmap of the newly packed zip, and then modify the latest-mac.yml file. After that, you will just need to manually upload the .zip and the .yml file to your release server.

**Comment by franzwarning** [comment]
Would love to see a fix for this. Experiencing it as well

**Comment by bcbee** [comment]
This is breaking our app as well, would love to see a fix 😀

**Comment by MehediH** [comment]
@bcbee if you use @wavifu's patch, it works for now. Obviously not the best fix, but did the job for me now. It's better than having updates blocked :)

**Comment by adam-lynch** [comment]
Another workaround is to create a second ZIP. Edit [out/targets/ArchiveTarget.js](https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/targets/ArchiveTarget.ts#L62) in node_modules. Duplicate the line where it creates the ZIP, so it'll create a second ZIP with a slightly different name.

```diff
+      await (0, _archive().archive)(format, artifactPath.replace('-mac.zip', '-mac-download.zip'), dirToArchive, archiveOptions);
       await (0, _archive().archive)(format, artifactPath, dirToArchive, archiveOptions);
```

Right after this, it appends the blockmap data to the auto-update ZIP. This won't be appended to the new (download) ZIP. Make sure this new download is uploaded to wherever you store downloads.

It's possible to name them `...-mac-update.zip` (auto-update ZIP) and `...-mac.zip` (download ZIP) if you play around with the electron-builder config too.

**Comment by harshitsilly** [comment]
Workaround for this is to use this gist.
https://gist.github.com/harshitsilly/a1bd5a405f93966aad20358ae6c4cec5
It will make the zip again with the use of ditto and generate the blockmap file(check mac-differential update - https://github.com/imjsElectron/electron-differential-updater) and its data and update in latest-mac.yml

**Comment by aabuhijleh** [comment]
I'm using the `dmg` target for distributing my app. I'm only using the `zip` target because `electron-updater` needs it for updates to work.

After testing, this issue seems to be not affecting automatic updates. Is that accurate? Or did anyone find any issues here?

**Comment by cenkalti** [comment]
In my case, I instruct users to install the app from DMG. Zip does not work when installing the app but it is required for automated updates to work.

**Comment by aabuhijleh** [comment]
> In my case, I instruct users to install the app from DMG. Zip does not work when installing the app but it is required for automated updates to work.

Yes, this is what I'm doing as well.

My question is:

Does this issue "The application xyz can't be opened" affect automatic updates?

For me, the answer seems to be: No

But I want to make sure

**Comment by cenkalti** [comment]
> Does this issue "The application xyz can't be opened" affect automatic updates?

My answer is no as well.

**Comment by mikekreeki** [comment]
Confirming whats said above, installation from zip does not work but it's required for automatic updates for which it works 👍

**Comment by paragbaxi** [comment]
Same issue:
./dist/mac/app execution blocked
./dist/app.dmg execution works

**Comment by Ryuurock** [comment]
Same issue: `The application "app" cannot be opened.`

**Comment by jthrilly** [comment]
Yes this is still relevant. There are workarounds, but the issue needs a PR to resolve it in electron-builder.

**Comment by jookyboi** [comment]
Just tried it again on version `22.9.1` of `electron-builder`. Still the same problem.

**Comment by harshitsilly** [comment]
Try this gist https://gist.github.com/harshitsilly/a1bd5a405f93966aad20358ae6c4cec5

**Comment by hovancik** [comment]
Yes.

**Comment by mmaietta** [comment]
Let me see if I can give this a shot. I see a few different options here:

1. Generate 2 zips, one for electron-updater and another for direct downloads. (https://github.com/electron-userland/electron-builder/issues/4299#issuecomment-654106344)
2. Disable embedding blockmap size for mac to avoid modifying the zip after compression (https://github.com/electron-userland/electron-builder/issues/4299#issuecomment-541249788)
3. Replace 7zip to use ditto on mac?

When using any of these custom post-EB scripts (`ditto <>.app` -> rerun `blockmap` -> update `latest-mac.yml`), is that zip able to be used for both electron-updater and direct-download/install?

AFAICT, this seems more of a limitation imposed upon by MacOS, similar as to how we can use `gatekeeperAssess: true` in our EB config. So we could approach with another flag for `disableEmbeddedBlockmap` or something akin.

I took a look at the EB source and I can't identify what `app-builder-bin blockmap {...}` is doing that we can achieve all in one pass, as opposed to "appending" and modifying the zip signature. I don't use electron-updater, so I'm very unfamiliar with how the embedded blockmap is utilized specifically on mac.

Happy to put together a PR, just not sure on what route to go.

**Comment by vitor251093** [comment]
It is still relevant.

**Comment by vitor251093** [comment]
Also, according to AWstats from cPanel, it seems that the automatic update system is downloading both the zip and the dmg for most of my users. I assume it's downloading the zip, noticing it was broken, and then downloading the dmg to solve the problem. I've recently made the dmg the default download (which is what I assume the path and sha512 of the YML root is), so I expect to put that theory to the test soon.

**Comment by fmiceli24** [comment]
Still relevant!!!! I have to edit `node_modules/app-builder-lib/out/targets/ArchiveTarget.js` as explained below just to have a zip file that makes the updater work.

> **Update:** Found the culprit: https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/targets/ArchiveTarget.ts#L65
> 
> After creating the zip, the default behavior for `ArchiveTarget` is to append the `.blockmap` to the file, then calculate the `blockmapsize` from that. This is in the name of doing differential updates.. though I've never seen `electron-updater` update this way for macOS.
> 
> I dug into `node_modules/app-builder-lib/out/targets/ArchiveTarget.js` and commented out this functionality:
> 
> ```
> if (this.isWriteUpdateInfo && format === "zip") {
>   // updateInfo = await (0, _differentialUpdateInfoBuilder().appendBlockmap)(artifactPath);
> }
> ```
> 
> The resulting `latest-mac.yml` will be missing the `blockmapsize` attribute but it doesn't seem to matter for me.
> 
> The resulting zip works fine on Catalina as does the auto-update functionality.
> 
> My theory is that by altering the zip file in this way (after compressing the notarized `.app/` folder), we're confusing Catalina's Gatekeeper functionality.

**Comment by mmaietta** [comment]
@fmiceli24 you can use this with `patch-package` to have it persist for MacOS builds specifically. I'm not sure what the implications of the blockmap not being present will be for electron-updater.
If it is indeed safe to disable blockmaps for mac zip target, then it's easy to turn this into a PR

`app-builder-lib+22.11.5.patch`
```
diff --git a/node_modules/app-builder-lib/out/targets/ArchiveTarget.js b/node_modules/app-builder-lib/out/targets/ArchiveTarget.js
index 77bdfb0..b788d5a 100644
--- a/node_modules/app-builder-lib/out/targets/ArchiveTarget.js
+++ b/node_modules/app-builder-lib/out/targets/ArchiveTarget.js
@@ -59,7 +59,8 @@ class ArchiveTarget extends core_1.Target {
                 withoutDir,
             };
             await archive_1.archive(format, artifactPath, dirToArchive, archiveOptions);
-            if (this.isWriteUpdateInfo && format === "zip") {
+            // Disabled for mac zips: https://github.com/electron-userland/electron-builder/issues/4299
+            if (this.isWriteUpdateInfo && format === "zip" && !isMac) {
                 updateInfo = await differentialUpdateInfoBuilder_1.appendBlockmap(artifactPath);
             }
         }
```

**Comment by vitor251093** [comment]
Thanks to @jookyboi 's 7za command, I've managed to replace the ZIP file with a functional one, but latest-mac-yml still had the info from the original ZIP file, so I had to create the JS file below, which receives the app version as an argument to work:

```
const fs = require('fs')
const yaml = require('yaml')
const child_process = require('child_process')

var args = process.argv.slice(2);
var version = args[0]

var filePath = './dist/latest-mac.yml'
var dataYaml = fs.readFileSync(filePath, {encoding: 'utf-8'})
var data = yaml.parse(dataYaml)

var zipFilePath = "./dist/My App-{version}-mac.zip".replace("{version}",version)
var regCommand = "shasum -a512 " + zipFilePath.replace(" ", "\\ ")
var output = child_process.execSync(regCommand, {encoding: 'utf-8'})
output = output.split(" ").shift().split("\t").shift()
output = new Buffer(output, 'hex').toString('base64') 
var stats = fs.statSync(zipFilePath)

var zipFile = data.files.filter(e => e.url.endsWith(".zip")).pop()
zipFile.sha512 = output
zipFile.size = stats.size

var dmgFile = data.files.filter(e => e.url.endsWith(".dmg")).pop()
data.path = dmgFile.url
data.sha512 = dmgFile.sha512

fs.writeFile(filePath, yaml.stringify(data), 'utf-8', function(){});
```

This is definitively just a workaround, and it requires `yaml`, but at least it is a workaround that worked for me, and that may work for someone else, so I decided to post it here.

**Comment by fmiceli24** [comment]
> diff --git a/node_modules/app-builder-lib/out/targets/ArchiveTarget.js b/node_modules/app-builder-lib/out/targets/ArchiveTarget.js
> index 77bdfb0..b788d5a 100644
> --- a/node_modules/app-builder-lib/out/targets/ArchiveTarget.js
> +++ b/node_modules/app-builder-lib/out/targets/ArchiveTarget.js
> @@ -59,7 +59,8 @@ class ArchiveTarget extends core_1.Target {
>                  withoutDir,
>              };
>              await archive_1.archive(format, artifactPath, dirToArchive, archiveOptions);
> -            if (this.isWriteUpdateInfo && format === "zip") {
> +            // Disabled for mac zips: https://github.com/electron-userland/electron-builder/issues/4299
> +            if (this.isWriteUpdateInfo && format === "zip" && !isMac) {
>                  updateInfo = await differentialUpdateInfoBuilder_1.appendBlockmap(artifactPath);
>              }
>          }

Thanks. Did that in order to have it streamlined. I am not using the blockmap. I understand this is for delta updates, right?

If delta updates are not working on OSX, and enabling the blockmap breaks updates altogether, would it be possible to have this workaround implemented upstream in order to avoid patching?

**Comment by sourabhverificient** [comment]
i solved this problem by using different target packaging in electron-builder "tar.xz"

 "mac": {
    "target": ["tar.xz"]
  },

**Comment by gniezen** [comment]
> i **### solved** this problem by using different target packaging in **electron-builder "tar.xz"**

Correct me if I'm wrong, but doesn't Squirrel.Mac require .zip artefacts? See (2) in Note here: https://www.electron.build/auto-update#quick-setup-guide

**Comment by sourabhverificient** [comment]
> > i **### solved** this problem by using different target packaging in **electron-builder "tar.xz"**
> 
> Correct me if I'm wrong, but doesn't Squirrel.Mac require .zip artefacts? See (2) in Note here: https://www.electron.build/auto-update#quick-setup-guide

your right , in my use case i just wanted to compressed app which can launch after unzip i am not planning to **autoupdate** it at this moment.

**Comment by RobertLowe** [comment]
After a lot of pain I finally got this working, I spent more time figuring out codesign-ing and this issue than I did developing my app.

You can see my implementation here: https://github.com/RobertLowe/force-audio which is based on https://gist.github.com/harshitsilly/a1bd5a405f93966aad20358ae6c4cec5

I use a two step process:

1. build the release as electron-builder wants, 
2. run the patcher AND reupload (and overwrite) to github releases.

**Comment by mmaietta** [comment]
I discovered that I can just swap it to generate an external `.blockmap` file so the zip isn't modified. Opened PR #6023

**Comment by 0xdeafcafe** [comment]
@mmaietta Just used your changes in my local node_modules folder and this does work perfectly 🎉, thank you so much! Do you know when the change will make it over to npm?

**Comment by mmaietta** [comment]
@0xdeafcafe I've just published `next` v22.11.9 to npm 🙂 
https://github.com/electron-userland/electron-builder/releases/tag/v22.11.9

Please note though...this issue was just reported but I've not been able to repro it in my 2 projects. So that's why I've published it as `next`
https://github.com/electron-userland/electron-builder/issues/6045
If anyone could confirm if they experience the issue, please comment on that ticket and any additional debug data would be immensely helpful.

**Comment by Vishal1419** [comment]
Writing this comment to help someone else facing the same issue in future.

I am using MacOS Mojave for development purposes
I was using electron-builder v22.9.x and creating a zip file
Then I was transferring generated .zip to my new laptop with MacOS Monterey via slack
Then I was using Archive Utility to unzip the app. (If I use The Unarchiever it worked fine)
I always got the error: Application cannot be opened

But when I saw above message from @mmaietta, I updated electron builder to v22.13.1 and then repeated the same process
Now, I can see that blockmap file is generated correctly and my app on MacOS Monterey opens up correctly.

Thank you @mmaietta for resolving the issue.
